JWBestPracticeApps/Casting/CastingViewController.swift

```diff
@@ -0,0 +1,185 @@
+    
+    var castingButton: UIButton? = nil
+    var barButtonItem: UIBarButtonItem? = nil
+    
+    var casting = false {
+        didSet {
+            castingButton?.tintColor = casting ? UIColor.green : UIColor.blue
+        }
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.navigationController?.navigationBar.barTintColor = UIColor.gray
+        
+        // Setup JWCastController object
+        setupCastController()
+    }
+    
+    func setupCastController() {
+        guard let player = self.player else { return }
+        
+        let castController = JWCastController(player: player)
+        castController.chromeCastReceiverAppID = kGCKDefaultMediaReceiverApplicationID
+        castController.delegate = self
+        castController.scanForDevices()
+        self.castController = castController
+    }
+    
+    func setupCastingButton() {
+        let buttonFrame = CGRect(x: 0, y: 0, width: 22, height: 22)
+        let castingButton = UIButton(frame: buttonFrame)
+        castingButton.addTarget(self, action: #selector(castButtonTapped(sender:)), for: .touchUpInside)
+        prepareCastingButtonWithAnimation(castingButton)
+        
+        let barButtonItem = UIBarButtonItem(customView: castingButton)
+        self.navigationItem.rightBarButtonItem = barButtonItem
+        
+        castingButton.heightAnchor.constraint(equalToConstant: 22).isActive = true
+        castingButton.widthAnchor.constraint(equalToConstant: 22).isActive = true
+        
+        self.castingButton = castingButton
+        self.barButtonItem = barButtonItem
+        
+        setCastingStatus(false)
+    }
+    
+    func prepareCastingButtonWithAnimation(_ button: UIButton) {
+        let connectingImages = [UIImage(named: "cast_connecting0")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting1")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting2")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting1")?.withRenderingMode(.alwaysTemplate)]
+        // Compact map to avoid nil UIImage objects
+        button.imageView?.animationImages = connectingImages.compactMap {$0}
+        button.imageView?.animationDuration = 2
+    }
+    
+    
+// pragma MARK: - Casting Status Helpers
+    
+    func startConnectingAnimation() {
+        castingButton?.tintColor = UIColor.white
+        castingButton?.imageView?.startAnimating()
+    }
+    
+    func stopConnectingAnimation(connected: Bool) {
+        castingButton?.imageView?.stopAnimating()
+        let castingImage = connected ? "cast_on" : "cast_off"
+        castingButton?.setImage(UIImage(named: castingImage)?.withRenderingMode(.alwaysTemplate), for: .normal)
+        castingButton?.tintColor = UIColor.blue
+    }
+    
+    func setCastingStatus(_ casting: Bool) {
```

Comment:
Because the rest of the code has been moved into the didSet, you no longer need this method. Instead of calling it, wherever the casting property needs to be changed, you can explicitly set it (e.g. self.casting = true)